### PR TITLE
[vlm] feat: support GLM-5.2 and Kimi K2.6 multimodal LoRA

### PR DESCRIPTION
**Pull request by dyedd**
## Summary

  Add GLM-5.2 and Kimi K2.6 multimodal VLM support, including MoonViT integration, foundation LoRA fine-tuning, DSA IndexShare, and checkpoint conversion workflows.

  ## Changes

  - Add the GLM-5.2 78-layer configuration with 21 indexer-owning layers and 57 shared layers.
  - Integrate IndexShare into fused DSA and validate PP/VPP stage boundaries.
  - Add GLM-5.2 + Kimi K2.6 VLM and LoRA configurations with configurable foundation LoRA targets.
  - Register the GLM-5.2 HF chat template and support Kimi media processing with GLM image markers.
  - Add HF/MCore BF16 and FP8 checkpoint conversion scripts.
  - Add an atomic GLM/Kimi HF checkpoint composer with symlink and materialized-copy modes.
  - Improve checkpoint name-map handling, aliased tensor saving, strict-load fallback, and validation dataset construction.
  - Add focused tests for chat templates, checkpoint composition, conversion config, checkpoint loading, and PatchMerger behavior.

  ## Notes

  - IndexShare is incompatible with `--overlap-moe-expert-parallel-comm`.
  - Each PP/VPP chunk must start on an IndexShare computing layer.
  - The checkpoint composer uses symlinks by default; use `--materialize` for independent output files.
  - Documentation-only source changes were omitted because the target fork uses a different documentation layout.
  - Align the validation-dataset test expectation with the current `get_val_dataset()` fallback behavior before merge.

### CODE REVIEW

File: loongforge/data/multimodal/kimi_task_encoder.py [hunk 1]
```diff
@@ -418,6 +418,29 @@ def process_sft_vqa(self, context, answer, image):
         Returns:
             input_ids, target, attn_mask, imgs, image_grid_thw
         """
+        if isinstance(self.chat_template, HFChatTemplate):
```
Comment (kaimo455) [inline loongforge/data/multimodal/kimi_task_encoder.py:426]: **[blocking] Avoid re-expanding plugin-generated media tokens**

This branch also matches the existing `kimi-k2.6-hf` template. For that template, `KimiK25Plugin.process_messages()` already replaces one `<image>` placeholder with N `<|media_content|>` tokens. `process_sft_qa()` then unconditionally calls `_expand_media_content_tokens()`, which expands the first token into N and retains the remaining N−1, producing 2N−1 tokens for N image features.

Image VQA therefore fails the token/feature-count invariant for N > 1. Please make expansion a no-op when the plugin has already emitted the expected token count, or scope this route to GLM as a temporary fix. Please also add a non-mocked `kimi-k2.6-hf` VQA regression test that verifies the final token and feature counts.

**Review by kaimo455 — CHANGES_REQUESTED** [review]

File: configs/models/glm5.2/glm5_2.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# IndexShare (HF index_topk_freq / index_skip_topk_offset): computing layers are
+# [0, 1, 2, 6, 10, ..., 74] 0-indexed, i.e. 21 of 78 own an indexer.
+dsa_indexer_topk_freq: 4
+dsa_indexer_skip_topk_offset: 3
+# HF indexer_rope_interleave: true -- the indexer uses interleaved RoPE, unlike DeepSeek-V3.2.
+dsa_indexer_rope_interleaved: true
+# GLM-5.x does not Hadamard-rotate the indexer q/k (DeepSeek-V3.2 does).
+dsa_indexer_rotate_activation: false
+# HF nn.LayerNorm(index_head_dim, eps=1e-6); differs from the model-level norm epsilon.
+dsa_indexer_k_norm_epsilon: 1e-6
```
Comment (kaimo455) [inline configs/models/glm5.2/glm5_2.yaml:45]: **[blocking] Wire this epsilon into the fused DSA indexer**

The shipped multimodal LoRA recipe passes `--norm-epsilon 1e-5` and enables `--use-dsa-fused`, but the fused indexer still constructs `k_norm` with `eps=self.config.layernorm_epsilon`. This new `dsa_indexer_k_norm_epsilon: 1e-6` value is therefore ignored on the advertised path, changing normalized keys, index scores, and potentially the selected top-k tokens.

Please construct the fused `k_norm` with `getattr(self.config, "dsa_indexer_k_norm_epsilon", self.config.layernorm_epsilon)` and add a fused-path regression test that verifies the configured epsilon is used, ideally including numerical parity with portable DSA.

File: examples/glm5.2_vit/checkpoint_convert/compose_glm5_2_kimi_k2_6_hf.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    if output.exists():
+        raise FileExistsError(f"Output already exists: {output}")
+    output.parent.mkdir(parents=True, exist_ok=True)
+
+    glm_map = _load_weight_map(glm_hf)
+    kimi_map = {
+        key: shard
+        for key, shard in _load_weight_map(kimi_hf).items()
+        if key.startswith(KIMI_PREFIX)
+    }
```
Comment (kaimo455) [inline examples/glm5.2_vit/checkpoint_convert/compose_glm5_2_kimi_k2_6_hf.py:74]: **[blocking] Do not copy excluded Kimi language tensors with a selected vision shard**

Filtering `kimi_map` changes only the generated index; `_link_weight_map()` still links or copies each selected safetensors file in full.

For example, suppose one Kimi shard contains both:

- `vision_tower.patch_embed.weight`
- `model.embed_tokens.weight`

The composer keeps only the `vision_tower.*` entry in the output index, but the copied `kimi-*.safetensors` file still physically contains Kimi's `model.embed_tokens.weight`. The composed index may therefore map GLM's `model.embed_tokens.weight` to a GLM shard and the vision tensor to this Kimi shard, while both physical files contain the same language key.

The current loader uses the index only to select shard filenames, then `load_file()` reads every tensor from each selected shard and `state_dict.update()` merges them without filtering against the index. If the Kimi shard is merged last, its unindexed `model.embed_tokens.weight` overwrites the GLM value; if their shapes differ, loading may instead fail later. The shard order is also derived from a set, so the collision result is not deterministic.

Please rewrite selected Kimi shards to contain only the indexed `vision_tower.*` tensors, or enforce the composed weight-map filter in the loader. Add an integration test where a Kimi shard contains both `vision_tower.patch_embed.weight` and a conflicting `model.embed_tokens.weight`, then load the composed checkpoint through the real loader and verify that the GLM embedding is preserved.

**Review by kaimo455 — CHANGES_REQUESTED** [review]

**Review by kaimo455 — APPROVED** [review]
